src/coroutine/memoset/mod.rs

```diff
@@ -539,25 +786,18 @@ impl<F: LurkField, Q: Query<F>> Scope<Q, LogMemo<F>> {
 }
 
 impl<F: LurkField> CircuitScope<F, LogMemoCircuit<F>> {
-    fn from_queries<CS: ConstraintSystem<F>>(
+    fn new<CS: ConstraintSystem<F>>(
         cs: &mut CS,
         g: &mut GlobalAllocator<F>,
         s: &Store<F>,
         memoset: LogMemoCircuit<F>,
-        queries: &HashMap<Ptr, Ptr>,
-        transcribe_internal_insertions: bool,
+        provenances: &HashMap<ZPtr<Tag, F>, ZPtr<Tag, F>>,
     ) -> Self {
-        let queries = queries
-            .iter()
-            .map(|(k, v)| (s.hash_ptr(k), s.hash_ptr(v)))
-            .collect();
-
         Self {
             memoset,
-            queries,
+            provenances: provenances.clone(), // FIXME
```

Comment:
What's the issue? Just the cloning?

Reply:
Yeah. The whole end-to-end process involves more cloning of data structures than it should. Once we have a grasp of the entire process (including the not-yet-implemented chunking/planning/optimization phase), we should refactor so all bookkeeping happens in a way that lets us pass data to where it's eventually needed efficiently.